src/main/java/com/commercetools/sync/producttypes/utils/ProductTypeUpdateEnumActionsUtils.java

```diff
@@ -0,0 +1,236 @@
+     * are the enum instances.
+     *
+     * @param enumValues the list of enum values.
+     * @param <T>        the enum type of the elements of the list.
+     * @return a map with the enum value key as a key of the map, and the
+     *         enum value as a value of the map.
+     */
+    @Nonnull
+    public static <T extends WithKey> Map<String, T> getEnumValuesKeyMap(@Nonnull final List<T> enumValues) {
+        return enumValues
+            .stream()
+            .collect(toMap(WithKey::getKey, enumValue -> enumValue));
+    }
+
+    /**
+     * Given a list of new {@link EnumValue}s, gets a map where the keys are the enum value key, and the values
+     * are the enum instances.
+     *
+     * @param attributeDefinitionName the attribute definition name whose the enum values belong to.
+     * @param enumValues              the list of enum values.
+     * @param <T>                     the enum type of the elements of the list.
+     * @return a map with the enum value key as a key of the map, and the enum
+     *         value as a value of the map.
+     * @throws DuplicateKeyException in case there are enum values with duplicate keys.
+     */
+    @Nonnull
+    public static <T extends WithKey> Map<String, T> getEnumValuesKeyMapWithKeyValidation(
+        @Nonnull final String attributeDefinitionName,
+        @Nonnull final List<T> enumValues) throws DuplicateKeyException {
+
+        return enumValues.stream().collect(
+            toMap(WithKey::getKey, enumValue -> enumValue,
+                (enumValueA, enumValueB) -> {
+                    throw new DuplicateKeyException(format("Enum Values have duplicated keys. "
+                            + "Attribute definition name: '%s', Duplicated enum value: '%s'. "
+                            + "Enum Values are expected to be unique inside their attribute definition.",
+                        attributeDefinitionName, enumValueA.getKey()));
+                }
+            ));
+    }
+
+    /**
+     * Checks if there are any old enum values which are not existing in the {@code newEnumValues}.
+     * If there are, then "remove" enum values update actions are built.
+     * Otherwise, if there are no old enum values, then an empty list is returned.
+     *
+     * @param attributeDefinitionName the attribute definition name whose enum values belong to.
+     * @param oldEnumValues           the list of old enum values.
+     * @param newEnumValues           the list of new enum values.
+     * @param <T>                     the enum type of the elements of the list.
+     * @return a list of enum values update actions if there are old enum value
+     *         that should be removed.
+     *         Otherwise, if the enum values are identical, an empty optional is returned.
+     */
+    @Nonnull
+    public static <T extends WithKey> Optional<UpdateAction<ProductType>> buildRemoveEnumValuesUpdateActions(
```

Comment:
c00d98fc1077336d500cb9227a5ae8a7dd8d267d

Reply:
> Better use specific type RemoveEnumValues instead of Optional<UpdateAction<ProductType>>?

I can treat all update action as a more generic UpdateAction<ProductType> type, so then all update actions applied to the product type are the same and easier to manipulate in more abstract methods used at higher levels of the product type sync